### PR TITLE
Dispose emitters' disposables after terminal events

### LINKED ISSUE #401: Emitters' disposables are being disposed before terminal events
Introduced in #392, release `1.1.9`. To be fixed in `1.1.10`.

### PR DESCRIPTION
**Pull request by arkivanov**
Fixes #401

### CODE REVIEW

**Comment by arkivanov** [comment]
@CherryPerry  So the JS tests were randomly failing with ReferenceError because it does not like xxxByEmitter builders being inlined for some reason. As a workaround I split the impls for JS and non-JS (do not inline for JS). Will recheck on every Kotlin update. I think inline is important here due to benchmarks.

**Comment by arkivanov** [comment]
For some reason `detekt` started failing with `NoConsecutiveBlankLines` error. Had to fix.

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/CompletableByEmitter.kt [hunk 1]
```diff
@@ -1,35 +1,7 @@
...
-                }
-            }
-
-        observer.onSubscribe(emitter)
-        emitter.tryCatch { onSubscribe(emitter) }
-    }
+// Separate implementations are because JS tests are randomly failing with ReferenceError if the function is inlined.
+// So do not inline for JS at the moment.
+@Suppress("ForbiddenComment")
+// TODO: recheck later
```
Comment (CherryPerry) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/CompletableByEmitter.kt]: Create issue.

File: reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt [hunk 2]
```diff
@@ -31,10 +31,33 @@ actual open class DisposableWrapper actual constructor() : Disposable {
...
+    }
+
+    /**
+     * Atomically either replaces any existing [Disposable]
+     * with the specified one or disposes it if wrapper is already disposed.
+     * Does not dispose any replaced [Disposable].
+     *
+     * @param disposable a new [Disposable], will be disposed if wrapper is already dispose
+     * @return replaced [Disposable] if any
+     */
```
Comment (CherryPerry) [inline reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Remove doc? Expect already has it.

File: reaktive/src/nativeCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt [hunk 3]
```diff
@@ -26,23 +27,35 @@ actual open class DisposableWrapper actual constructor() : Disposable {
...
 
-    private fun setHolder(holder: Holder?) {
+    /**
+     * Atomically either replaces any existing [Disposable]
+     * with the specified one or disposes it if wrapper is already disposed.
+     * Does not dispose any replaced [Disposable].
+     *
+     * @param disposable a new [Disposable], will be disposed if wrapper is already dispose
+     * @return replaced [Disposable] if any
+     */
```
Comment (CherryPerry) [inline reaktive/src/nativeCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Same.

File: reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt (hunk 2 above)
Comment (arkivanov) [reply reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Unfortunately docs are not exported for actuals and ate missing then

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/CompletableByEmitter.kt (hunk 1 above)
Comment (arkivanov) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/CompletableByEmitter.kt]: Created #404

**Pull request merged**